### DISCUSSION TITLE
`z.enum()` breaks if you pass it an actual enum or array created by destructuring enum using Object.values()`

### DISCUSSION THREAD

**Discussion by rogadev** [colinhacks/zod#2125] [Q&A] [answered]
This can't possibly be how Zod is designed to handle enum validation...

I would have thought this would work:

```ts
import { z } from 'zod';

enum PizzaType {
  Canadian = 'Canadian',
  Cheese = 'Cheese',
  Hawaiian = 'Hawaiian',
  MeatLovers = 'Meat Lovers',
  Pepperoni = 'Pepperoni',
  Vegetarian = 'Vegetarian',
}

const orderSchema = z.object({
  id: z.number(),
  type: z.enum(PizzaType)
})

const order = {
  id: 1,
  type: 'Canadian'
}

orderSchema.safeParse(order)
```

Looks good to me. Logical. Simple. Should definitely work. Right?

```bash
error TS2769: No overload matches this call.
      Overload 1 of 2, '(values: readonly [string, ...string[]], params?: RawCreateParams): ZodEnum<[string, ...string[]]>', gave the following error.
        Argument of type 'typeof PizzaType' is not assignable to parameter of type 'readonly [string, ...string[]]'.
      Overload 2 of 2, '(values: [string, ...string[]], params?: RawCreateParams): ZodEnum<[string, ...string[]]>', gave the following error.
        Argument of type 'typeof PizzaType' is not assignable to parameter of type '[string, ...string[]]'.

    34     .enum(PizzaType)
                  ~~~~~~~
```

Okay, well that's weird. I guess I'll just do `Object.values(PizzaType)`?

```ts
const orderSchema = z.object({
  id: z.number(),
  type: z.enum(Object.values(PizzaType))
})
```

```bash
error TS2769: No overload matches this call.
      Overload 1 of 2, '(values: readonly [string, ...string[]], params?: RawCreateParams): ZodEnum<[string, ...string[]]>', gave the following error.
        Argument of type 'typeof PizzaType' is not assignable to parameter of type 'readonly [string, ...string[]]'.
      Overload 2 of 2, '(values: [string, ...string[]], params?: RawCreateParams): ZodEnum<[string, ...string[]]>', gave the following error.
        Argument of type 'typeof PizzaType' is not assignable to parameter of type '[string, ...string[]]'.

    34     .enum(Object.values(PizzaType))
                  ~~~~~~~~~~~~~~~~~
```

Uh, okay... How about destructuring into a new array of strings.

```ts
const orderSchema = z.object({
  id: z.number(),
  type: z.enum([...Object.values(PizzaType)])
})
```

```bash
error TS2769: No overload matches this call.
      Overload 1 of 2, '(values: readonly [string, ...string[]], params?: RawCreateParams): ZodEnum<[string, ...string[]]>', gave the following error.
        Argument of type 'typeof PizzaType' is not assignable to parameter of type 'readonly [string, ...string[]]'.
      Overload 2 of 2, '(values: [string, ...string[]], params?: RawCreateParams): ZodEnum<[string, ...string[]]>', gave the following error.
        Argument of type 'typeof PizzaType' is not assignable to parameter of type '[string, ...string[]]'.

    34     .enum([...Object.values(PizzaType)])
                  ~~~~~~~~~~~~~~~~~
```

Okay, well, I don't know how much uglier I need to get... I mean this would be even uglier, but I guess I could try:

```ts
const orderSchema = z.object({
  id: z.number(),
  type: z.enum([
      PizzaType.Canadian,
      PizzaType.Cheese,
      PizzaType.Hawaiian,
      PizzaType.MeatLovers,
      PizzaType.Pepperoni,
      PizzaType.Vegetarian,
    ]
})
```

```bash
✅ Yup! Looks good to me!
```

WHY, ZOD!? WHY!? 😭

**Comment by JacobWeisenburger** [comment, marked as answer]
Is this what you are looking for?
https://github.com/colinhacks/zod#native-enums

```ts
enum PizzaType {
    Canadian = 'Canadian',
    Cheese = 'Cheese',
    Hawaiian = 'Hawaiian',
    MeatLovers = 'Meat Lovers',
    Pepperoni = 'Pepperoni',
    Vegetarian = 'Vegetarian',
}

const orderSchema = z.object( {
    id: z.number(),
    type: z.nativeEnum( PizzaType )
} )

const order = {
    id: 1,
    type: 'Canadian'
}

orderSchema.safeParse( order )
```

Other Resources:
https://gist.github.com/JacobWeisenburger/1c72eda9e823cd6caed327db470f0613

If you found my answer satisfactory, please consider supporting me. Even a small amount is greatly appreciated. Thanks friend! 🙏
https://github.com/sponsors/JacobWeisenburger

**Reply by silentsiri** [reply]
but how about `const enum`

**Reply by fnandogp** [reply]
I usually create enum with `as const`.
```
// types/title.ts
import { getOptions } from '@lib/enum'

export const Title = {
  Mr: 'mr',
  Mrs: 'mrs',
  Miss: 'miss',
  Ms: 'ms',
  Dr: 'dr',
} as const
export type Title = (typeof Title)[keyof typeof Title]
```

Then I have this utility file to help me with zod.
```
// lib/enum.ts
export function getValues<T extends Record<string, any>>(obj: T) {
  return Object.values(obj) as [(typeof obj)[keyof T]]
}
```

When creating the zod object I use:
```
// schemas/profile.ts
export const SaveContactDetailsFields = z.object({
  title: z.enum(getValues(Title), {
    errorMap: () => ({
      message: 'Please select your title',
    }),
  }),
...
})
```

**Comment by rogadev** [comment]
Would it not make more sense to have `.enum()` handle real enums and instead have a `.enumString()` or `.enumLike()` method instead? It seems pretty counter intuitive to me. Better yet would be if `.enum()` handled both native and non-native JavaScript enums.

**Reply by JacobWeisenburger** [reply]
I'm not sure what you mean, because as far as I know, JavaScript doesn't have native enums. TypeScript has native enums, which are converted into just a regular object at runtime.

---

Why you should not use TypeScript Enums:
- [TypeScript's Worst Feature (And How To Avoid It)](https://www.youtube.com/watch?v=Anu8vHXsavo)
- [Enums considered harmful](https://www.youtube.com/watch?v=jjMbPt_H3RQ)
- [How to use TypeScript Enums and why not to, maybe](https://www.youtube.com/watch?v=pWPClHdcvVg)

**Reply by rogadev** [reply]
Sorry, mistyped it - TypeScript enums.

**Comment by mrhyde** [comment]
> Why you should not use TypeScript Enums:
> 
> * [TypeScript's Worst Feature (And How To Avoid It)](https://www.youtube.com/watch?v=Anu8vHXsavo)
> * [Enums considered harmful](https://www.youtube.com/watch?v=jjMbPt_H3RQ)
> * [How to use TypeScript Enums and why not to, maybe](https://www.youtube.com/watch?v=pWPClHdcvVg)

Before making a bold statement based on videos that try to attract attention with catchy titles, consider at least reading the comments. You will notice that many people disagree with the downsides described in those videos because they are often exaggerated or pulled out of thin air.

**Reply by jtmueller** [reply]
Also, if an argument isn't important enough to write down, it can't be a very good argument. I have no interest in spending 15 minutes watching a video to get information I could read in 5 minutes. If the only supporting information you have is in YouTube videos, you don't have supporting information.

**Reply by Zach-Jaensch** [reply]
Even TypeScript has started recommending not using them. The TS Enum docs suggest using JS Objects https://www.typescriptlang.org/docs/handbook/enums.html#objects-vs-enums

**Comment by mjp0** [comment]
So, let me get this straight: I can do `z.enum(["foo", bar"])` but I can't do `const e = ["foo", bar"]; z.enum(e)`?

This makes zero sense. I have a bunch of SSOT JSON lists which I though I could just push as enums.

**Reply by lorenzoa7** [reply]
Same issue here...

**Reply by andredezzy** [reply]
Same issue

**Reply by mariuslundgard** [reply]
Try `const e = ["foo", bar"] as const; z.enum(e)`

**Reply by scotttrinh** [reply]
The fundamental issue is the `z.enum` (as well as a few other Zod schema types) is written in a way that guarantees that the array you feed it has at least one value of the array's element type. This is the only way to actually get compile-time information about the _values_ of the enum.

If all you care about is having a schema that validates if a value is included in another array, you can make a custom Zod type like:

```ts
const e = ["foo", "bar"];

const oneOfE = z.custom<string>((val) => e.includes(val));
```

If, on the other hand you _do_ want to have access to `"foo"` and `"bar"` at compile time, and those strings are actually constant in your code and compile time (via assignment to a variable), you can use `as const` to indicate to TypeScript that this is a constant value in your program.

If it's actually dynamic (like in the example given of loading JSON from disk or the network, etc), then yeah, the `z.custom` check will work, but you won't get anything useful out of the type, just `string`, which might be fine for your use case!

**Comment by letstri** [comment]
I've created a utility function to enhance the type safety of Zod enums.

The `zodEnum` function takes an array of values and casts it to a tuple type where the first element is of the same type as the array, and the rest are of the same type as well.

```ts
const zodEnum = <T>(arr: T[]): [T, ...T[]] => arr as [T, ...T[]];
```

Now, you can use this function to satisfy Zod's enum typization.

For example, let's say you have a type 'UserType' and array with it:

```ts
type UserType = 'admin' | 'user';

const userTypes = ['admin', 'user'] satisfies UserType[];
```

By applying the `zodEnum` function, you satisfy Zod's enum expectations, and the type checker will no longer raise an error.

```ts
// With error
z.enum(userTypes);

// Without error
z.enum(zodEnum(userTypes));
```

Or you can validate it:

```ts
z.enum(zodEnum<UserType>(['admin', 'user']));
```

**Reply by allan-cannon** [reply]
I think there's an easier way to do this by including `as const`.

```typescript
type UserType = 'admin' | 'user';

// This is now a tuple
const userTypes = ['admin', 'user'] as const satisfies UserType[];

// Works as expected
z.enum(usertypes);
```

**Comment by babpulss** [comment]
I've tried like below 
z.enum(['', ...Object.values(obj)])

**Reply by AbubakarWebDev** [reply]
Maybe this could be considered as a hacky way to do it, but it works 100%, and you can also do it using the below approach:

```javascript
const ASPECT_RATIOS = {
  ONE_BY_ONE: '1:1',

  THREE_BY_TWO: '3:2',
  TWO_BY_THREE: '2:3',

  FOUR_BY_FIVE: '4:5',
  FIVE_BY_FOUR: '5:4',

  FOUR_BY_THREE: '4:3',
  THREE_BY_FOUR: '3:4',

  NINE_BY_SIXTEEN: '9:16',
  SIXTEEN_BY_NINE: '16:9',

  FIVE_BY_TWELVE: '5:12',
  TWELVE_BY_FIVE: '12:5',
} as const;

export const AspectRatioSchema = z.enum([
  ASPECT_RATIOS.ONE_BY_ONE,
  ...Object.values(ASPECT_RATIOS),
]);

// infer the above schema type
export type AspectRatioSchemaType = z.infer<typeof AspectRatioSchema>;
```

You can also verify the typescript inter for this type from the screenshot below:

<img width="1129" alt="Screenshot 2024-06-12 at 8 21 21 AM" src="https://github.com/colinhacks/zod/assets/61361764/268fdc55-b634-4e8b-a9c8-7dd393a80c0e">

The reason for the above solution working is that the Zod enum can accept read-only strings while spreading the read-only string array.

<img width="1132" alt="image" src="https://github.com/colinhacks/zod/assets/61361764/ae95e841-b295-4f66-b8b9-a8f00ca291d5">

**Comment by MendyLanda** [comment]
For anyone looking for a simple solution, just use this simple util
```ts
export const typedEnum = <T extends string>(arr: T[]) =>
  z.enum(arr as [T, ...T[]]);
```

yw

**Reply by MendyLanda** [reply]
p.s. you loose the safety of checking that there's at least one value in the enum. 
so maybe name this unsafeEnum and make sure you pass correct values.

**Comment by cupofjoakim** [comment]
To add to the conversation, `nativeEnum` has now been deprecated and passing a typescript enum to `z.enum()` is supported from Zod 4: https://zod.dev/v4/changelog?id=znativeenum-deprecated
